numpyro/contrib/module.py

```diff
@@ -433,3 +435,143 @@ def random_haiku_module(
         _update_params(params, new_params, prior)
     nn_new = partial(nn.func, new_params, *nn.args[1:], **nn.keywords)
     return nn_new
+
+
+def nnx_module(name, nn_module):
+    """
+    Declare a :mod:`~flax.nnx` style neural network inside a
+    model so that its parameters are registered for optimization via
+    :func:`~numpyro.primitives.param` statements.
+
+    Given a flax NNX ``nn_module``, to evaluate the module, we directly call it.
+    In a NumPyro model, the pattern will be::
+
+        # Eager initialization outside the model
+        module = nn_module(...)
+
+        # Inside the model
+        net = nnx_module("net", module)
+        y = net(x)
+
+    :param str name: name of the module to be registered.
+    :param flax.nnx.Module nn_module: a pre-initialized `flax nnx` Module instance.
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        from flax import nnx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use flax.nnx to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `flax` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/google/flax.git`."
+        ) from e
+
+    module_key = name + "$params"
+    module_params = numpyro.param(module_key)
+
+    if module_params is None:
+        graph_def, params_state, other_state = nnx.split(
+            nn_module, nnx.Param, nnx.filterlib.to_predicate(nnx.Not(nnx.Param))
+        )
+
+        params = nnx.to_pure_dict(params_state)
+
+        if params:
+            module_params = numpyro.param(module_key, params)
+
+        state_dict = {"state": params_state}
+
+        if any(state_dict.values()):
+            _ = numpyro_mutable(name + "$state", state_dict)
+
+    def apply_fn(*call_args, **call_kwargs):
+        user_rngs = call_kwargs.pop("rngs", None)
+
+        model = nn_module
+
+        if module_params:
+            graph_def, state = nnx.split(model)
```

Comment:
you need to merge the latest params and mutable state here. Something like
```
if module_params:
    nnx.replace_by_pure_dict(state, module_params)
if state_dict:
    nnx.replace_by_pure_dict(state, state_dict)
```

Reply:
added in https://github.com/pyro-ppl/numpyro/pull/1990/commits/6cd66c3a607ac10b321dbd1c823255822eef8624

Reply:
I am unsure where this goes after https://github.com/pyro-ppl/numpyro/pull/1990/commits/048c253f7af4b5c52ea912ea55d8cb5ea3f26d14 🙏

Reply:
I think you can do
```
if params:
    nnx.replace_by_pure_dict(eager_params_state, params)
if mutable_holder:
    nnx.replace_by_pure_dict(eager_other_state, mutable_holder["state"])
model = nnx.merge(graph_def, eager_params_state, eager_other_state)
model_call = model(*call_args, **call_kwargs)
_, new_mutable_state = nnx.split(model, nnx.filterlib.to_predicate(nnx.Not(nnx.Param)))
if mutable_holder:
    mutable_holder["state"] = new_mutable_state
```